src/datasets/table.py

```diff
@@ -1744,6 +1799,75 @@ def wrapper(array, *args, **kwargs):
     return wrapper
 
 
+def array_concat(arrays: List[pa.Array]):
+    """Improved version of pa.concat_arrays
+
+    It supports concatenating pa.ExtensionArray objects by concatenating the underlying storages.
+
+    Args:
+        arrays (List[pa.Array]): List of arrays to contatenate
+
+    Raises:
+        pa.ArrowInvalid: if the arrow array concatenation fails
+        ValueError: if the list of arrays is empty
+        TypeError: if the arrays to be concatenated have different types
+
+    Returns:
+        array (:obj:`pyarrow.Array`): the concatenated array
+    """
+    arrays = list(arrays)
+    array_types = set(array.type for array in arrays)
+
+    if not array_types:
+        raise ValueError("Couldn't concatenate empty list of arrays")
+    if len(array_types) > 1:
+        array_types = list(array_types)
+        raise TypeError(f"Couldn't concatenate arrays with different types {array_types[0]} and {array_types[1]}")
+
+    def _offsets_concat(offsets):
+        offset = offsets[0]
+        concatenated_offsets = offset
+        for offset in offsets[1:]:
+            offset = pc.subtract(offset, offset[0])
+            offset = pc.add(offset[1:], concatenated_offsets[-1])
+            concatenated_offsets = pa.concat_arrays([concatenated_offsets, offset])
+        return concatenated_offsets
+
+    def _concat_arrays(arrays):
+        array = arrays[0]
+        if isinstance(array, pa.ExtensionArray):
+            return array.type.wrap_array(_concat_arrays([array.storage for array in arrays]))
+        elif pa.types.is_struct(array.type):
+            return pa.StructArray.from_arrays(
+                [_concat_arrays([array.field(field.name) for array in arrays]) for field in array.type],
+                fields=list(array.type),
+                mask=pa.concat_arrays([array.is_null() for array in arrays]),
+            )
+        elif pa.types.is_list(array.type):
+            if config.PYARROW_VERSION.major < 10:
+                warnings.warn(
+                    "None values are converted to empty lists in `pyarrow<10.0.0` when concatenating list arrays with None values. Install `pyarrow>=10.0.0` to avoid this behavior. More info: https://github.com/huggingface/datasets/issues/3676."
+                )
```

Comment:
We may have a regression here: all the users with pyarrow<10 will experience their nulls converted to empty lists in ListArray when they run `.map()`.

Maybe we should use pa.concat_arrays when there's no extension arrays in a ListArray ?